lib/core/unittest/CProgramCountersTest.cc

```diff
@@ -0,0 +1,386 @@
+        inserter.toXml(newStaticsXml);
+    }
+
+    // we expect the persisted counters without first caching to be the
+    // same as those from when we do cache the live values,
+    // as persistence uses live values if the cache is not available
+    CPPUNIT_ASSERT_EQUAL(newStaticsXml, newStaticsXmlNoCaching);
+
+    // Restore the original all-zero state
+    {
+        ml::core::CRapidXmlParser parser;
+        CPPUNIT_ASSERT(parser.parseStringIgnoreCdata(origStaticsXml));
+        ml::core::CRapidXmlStateRestoreTraverser traverser(parser);
+        CPPUNIT_ASSERT(traverser.traverseSubLevel(
+            ml::core::CProgramCounters::staticsAcceptRestoreTraverser));
+    }
+
+    for (int i = 0; i < ml::counter_t::E_LastEnumCounter; i++) {
+        CPPUNIT_ASSERT_EQUAL(TCounter(0), counters.counter(i));
+    }
+
+    // Restore the non-zero state
+    {
+        ml::core::CRapidXmlParser parser;
+        CPPUNIT_ASSERT(parser.parseStringIgnoreCdata(newStaticsXml));
+        ml::core::CRapidXmlStateRestoreTraverser traverser(parser);
+        CPPUNIT_ASSERT(traverser.traverseSubLevel(
+            &ml::core::CProgramCounters::staticsAcceptRestoreTraverser));
+    }
+
+    for (int i = 0; i < ml::counter_t::E_LastEnumCounter; i++) {
+        CPPUNIT_ASSERT_EQUAL(TCounter(567 + (i * 3)), counters.counter(i));
+    }
+
+    // Restore the zero state to clean up
+    {
+        ml::core::CRapidXmlParser parser;
+        CPPUNIT_ASSERT(parser.parseStringIgnoreCdata(origStaticsXml));
+        ml::core::CRapidXmlStateRestoreTraverser traverser(parser);
+        CPPUNIT_ASSERT(traverser.traverseSubLevel(
+            &ml::core::CProgramCounters::staticsAcceptRestoreTraverser));
+    }
+
+    // Check that the cache is automatically cleaned up
+    CPPUNIT_ASSERT(counters.m_Cache.empty());
+
+    for (int i = 0; i < ml::counter_t::E_LastEnumCounter; i++) {
+        CPPUNIT_ASSERT_EQUAL(TCounter(0), counters.counter(i));
+    }
+
+    // check that the format of the output stream operator is as expected
+    std::ostringstream ss;
+    ss << counters;
+    const std::string output(ss.str());
+
+    ml::core::CRegex::TStrVec tokens;
+    {
+        ml::core::CRegex regex;
+        regex.init("\n");
+        regex.split(output, tokens);
+    }
+    for (ml::core::CRegex::TStrVecCItr i = tokens.begin(); i != (tokens.end() - 1); ++i) {
+        ml::core::CRegex regex;
+        // Look for "name":"E.*"value": 0}
+        regex.init(".*\"name\":\"E.*\"value\":0.*");
+        CPPUNIT_ASSERT(regex.matches(*i));
+    }
+
+    LOG_DEBUG(<< output);
+
+    // Test persistence/restoration of a subset  of counters
+    auto testCounterSubset = [&](const ml::counter_t::TCounterTypeSet& counterSet) {
+
+        // Register interest in a subset of counters
+        ml::core::CProgramCounters::registerProgramCounterTypes(counterSet);
+
+        // Check that a save/restore with all zeros is Ok
+        for (int i = 0; i < ml::counter_t::E_LastEnumCounter; i++) {
```

Comment:
nit: coding guidelines are we should always use prefix form unless you really need the returned value (and elsewhere).